### PR TITLE
Add regression tests for six silently-fixed issues (close #4317 #4859 #4860 #4861 #4863 #5154)

### LINKED ISSUE #4317: Starmapping too many times crashes rustpython.
In the following example, b is an empty list, we execute starmap() on b for 100000, then rustpython crashes with segmentation fault. This case can work well on CPython 3.9.1.

test.py:
```
from operator import add
from itertools import starmap
b=[]
for i in range(100000):
    b = starmap(i, zip(b, b))
list(b)
```

Behavior on rustpython(rustpython 0.1.2):
`Segmentation fault: 11`

Behavior on CPython 3.9.1(Expected):
work well on the python interpreter(work well on Python 3.9.1)

Step to reproduce:
type "cargo run --release test.py" on the console of Ubuntu 18.04 or MacOS Big Sur 11.7.1
rust version: rustc 1.65.0

### LINKED ISSUE #4859: The string representation of the asyncio object results in 'panicked'.
The following code defines an asynchronous function goo that takes two arguments a and b. The asyncio.create_task() function is then used inside the main function to create a task object that represents the execution of the goo coroutine with arguments 1 and b=1. Finally, the repr() function obtains a string representation of the task object. The asyncio.run() function runs the main coroutine and execute the task. The code essentially creates and runs an empty task while RustPython gets crashing.

```
import asyncio

async def goo(a, b): pass

async def main():
    task = asyncio.create_task(goo(1, b=1))
    print(repr(task))

asyncio.run(main())
```

Crash message: 
```
thread 'main' panicked at 'index out of bounds: the len is 2 but the index is 18446744073709551615', vm/src/frame.rs:169:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Environment:
Ubuntu 18.04
rustpython v0.2.0

### LINKED ISSUE #4860: Segmentation fault when frequently creating and removing references in XML tree
This code generates an XML tree structure with 20001 nested elements, where each element is named 'x'. Initially, a root element named 'x' is created using  xml.etree.cElementTree. Then, the for loop creates 20000 sub-elements of the root element. Each sub-element is also named 'x' and is created using the SubElement method. The variable y is assigned the last sub-element created. It is a reference to the deepest nested element in the tree. Finally, the variable x is deleted, which removes the reference to the root element. However, the y variable still holds the reference to the entire XML tree structure, including the root element and all of its sub-elements.

test.py
```
import xml.etree.cElementTree as ElementTree
y = x = ElementTree.Element('x')
for i in range(20000): 
    y = ElementTree.SubElement(y, 'x')
del x
```

Crash message:
Segmentation fault (core dumped)

Environment:
Ubuntu 18.04
rustpython v0.2.0

### LINKED ISSUE #4861: Null reference when a variable name of the deleted object is assigned the second time.
This code defines a class called C and creates an instance of this class named c. Inside the __del__ method, a new instance of the C class is created and assigned to a variable c. Then, the code creates a variable d which contains a range object from 0 to 9. Finally, the variable c is assigned to the value of d, which means that the original instance of the C class created earlier is no longer referenced and will be garbage collected, while rustpython fails doing that and get panicked.

```
class C:
    def __del__(self):
        c = C()
c = C()
d = range(10) 
c = d    #crash here
```

Crash message:
```
Exception ignored in: <object function repr() failed>
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: PyBaseException', vm/src/stdlib/sys.rs:593:59
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Environment:
Ubuntu 18.04
rustpython v0.2.0

### LINKED ISSUE #4863: Panicked if type(self)() in __del__ methods
This code creates an instance of the Dummy class and then immediately deletes it, which triggers the __del__ method to create a new instance of the Dummy class. 

```
class Dummy:
   def __del__(self):
      type(self)()

d = Dummy()
del d
```

Crash message:
```
Exception ignored in: <object function repr() failed>
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: PyBaseException', vm/src/stdlib/sys.rs:593:59
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Environment:
Ubuntu 18.04
rustpython v0.2.0

### LINKED ISSUE #5154: Crash about overflowing max_stackdepth
## Summary

I tried to run [scrapscript.py](https://github.com/tekknolagi/scrapscript) with RustPython. It crashed. I reduced the input code with CReduce and then manually from ~4k lines to 21 lines. 

## Expected

No crash. In the full example, I expect the test suite to run. In the reduced example I expect nothing to happen.

## Actual

```python
# repro.py
class LeftParen:
    pass
class RightParen:
    pass
class Lexer:
    def read_char(self) :
        return None
    def read_one(self) :
        while self:
            c = self.read_char()
            break
        else:
            return x
        {
                "(": LeftParen,
                ")": RightParen}
def tokenize():
    lexer = Lexer()
    while (lexer.read_one()) :
        pass
tokenize()
```

```
cedar% rustpython scrapscript.py test
[vm/src/frame.rs:1942] self = ExecutingFrame {
    code: code: <code object read_one at ??? file "scrapscript.py", line 8>,
    state: FrameState {
        stack: [
            [PyObject PyStr { value: "(", kind: Ascii, hash: -1 }],
            [PyObject [PyType LeftParen]],
            [PyObject PyStr { value: ")", kind: Ascii, hash: -1 }],
        ],
        blocks: [],
        lasti: 16,
    },
}
thread 'main' panicked at vm/src/frame.rs:1943:9:
tried to push value onto stack but overflowed max_stackdepth
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
cedar%
```

### PR DESCRIPTION
**Pull request by changjoon-park**
## Summary

Six GitHub issues report crashes / panics that the upstream code base has since fixed, but whose issues were never linked by a closing PR and remain open years later. This PR adds a single snippet file that pins each minimal reproduction so future refactors cannot regress any of them without a visible test failure, and closes the open issues via the PR body.

## Issues closed

| # | Original symptom | Current status |
|---|---|---|
| [#4317](https://github.com/RustPython/RustPython/issues/4317) | `starmap(i, zip(b, b))` repeated ~100k times on an empty list — SIGSEGV | Returns `[]` cleanly |
| [#4859](https://github.com/RustPython/RustPython/issues/4859) | `repr(asyncio.Task)` inside `asyncio.run` — Rust panic | Returns a non-empty string |
| [#4860](https://github.com/RustPython/RustPython/issues/4860) | Deeply-nested `xml.etree` SubElements — segfault on cleanup | Tree is freed normally |
| [#4861](https://github.com/RustPython/RustPython/issues/4861) | Reassigning a name while `__del__` runs — null reference panic | `__del__` fires; reassignment completes |
| [#4863](https://github.com/RustPython/RustPython/issues/4863) | `__del__` calling `type(self)()` — `Result::unwrap on Err` panic | `type(self)()` returns normally in a guarded `__del__` |
| [#5154](https://github.com/RustPython/RustPython/issues/5154) | Compiler crash on reduced `scrapscript.py` (while/else + trailing dict literal) | `compile()` accepts the source |

## Changes

A single new file: `extra_tests/snippets/regression_closed_issues.py` (+143 lines).

Each section has a comment block referencing the original issue, explaining what the old crash was, and running the minimal repro. Where the original report used an unbounded `__del__`-cascade or very large iteration count purely to surface the crash, the regression test uses the same shape with a bounded inner loop or one-shot guard — the bug class is exercised without depending on runtime cost at garbage-collection time.

## Verification

```console
$ ./target/release/rustpython extra_tests/snippets/regression_closed_issues.py
$ echo $?
0

$ python3 extra_tests/snippets/regression_closed_issues.py
$ echo $?
0

$ cargo fmt --check
```

Both interpreters pass immediately — no test depends on RustPython-specific timing or GC traversal cost.

## Scope

- **In**: regression coverage that locks in already-working behavior.
- **Out**: any functional code change. The file is test-only; no crate under `crates/` is touched.

## Why bundle as one PR

Six issues from the same class ("pre-existing SIGSEGV / panic, fixed implicitly, never closed") in one PR keeps reviewer surface small (one file, all test-only) and makes the intent unambiguous — the PR is an issue-tracker cleanup, not a feature. Similar to #7625 which unmasked two `@unittest.expectedFailure` markers in the same session that introduced the underlying fix.

## Summary by CodeRabbit

* **Tests**
  * Added a regression test suite that re-validates several previously fixed issues: self-referential iterator termination in corner cases; asyncio task creation, representation and awaited results; robust cleanup of deeply nested XML element trees during teardown; two destructor/finalization regression scenarios (flag-setting and guarded re-entry); and a compiler-stability check for a previously problematic source pattern.

Closes #4317.
Closes #4859.
Closes #4860.
Closes #4861.
Closes #4863.
Closes #5154.

### CODE REVIEW

File: extra_tests/snippets/regression_closed_issues.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+
+# --- #4317: starmap over zip of an empty list, repeated -------------------
+#
+# Original report: `b = []; for i in range(100000): b = starmap(i, zip(b, b));
+# list(b)` segfaulted. The iterator was self-referential and the Rust stack
+# unwound until SIGSEGV. Evaluation now terminates cleanly.
+_b = []
+for _i in range(1000):
+    _b = starmap(_i, zip(_b, _b))
```
Comment (youknowone) [inline extra_tests/snippets/regression_closed_issues.py:26]: this is not worth to add. we have starmap tests in `Lib/test/test_itertools.py`
Closing the test will be enoguh

File: extra_tests/snippets/regression_closed_issues.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+assert list(_b) == []
+
+
+# --- #4859: repr() of an asyncio Task -------------------------------------
+#
+# Original report: inside `asyncio.run`, constructing a Task and calling
```
Comment (youknowone) [inline extra_tests/snippets/regression_closed_issues.py:32]: `stdlib_asyncio.py` will be better

File: extra_tests/snippets/regression_closed_issues.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+
+# --- #4860: deep XML tree cleanup ----------------------------------------
+#
+# Original report: building ~20000 nested XML SubElements then dropping
+# references segfaulted during GC. Running the same pattern (smaller depth
+# here to keep the snippet fast) now completes cleanly when all element
+# references are released within the test — relying on module teardown
+# would let a cleanup-path regression hide from this snippet.
+_root = ET.Element("x")
```
Comment (youknowone) [inline extra_tests/snippets/regression_closed_issues.py:55]: I expect test_xml famliy covers this.

File: extra_tests/snippets/regression_closed_issues.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+
+# --- #4863: __del__ calling type(self)() ---------------------------------
+#
+# Original report: `class Dummy: def __del__(self): type(self)()` + `del d`
+# panicked with `Result::unwrap() on an Err`. The regression exercises the
+# `type(self)()` pattern inside `__del__` exactly once; a one-shot guard
+# prevents the constructor chain from cascading (which CPython handles via
+# its recursion limit but at noticeable cost).
+class _Dummy4863:
```
Comment (youknowone) [inline extra_tests/snippets/regression_closed_issues.py:93]: This must be covered by some Lib/test tests, but if not sure, `builtin_type.py` please.

File: extra_tests/snippets/regression_closed_issues.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            ")": RightParen,
+        }
+
+
+def tokenize():
+    lexer = Lexer()
+    while lexer.read_one():
+        pass
+'''
+compile(_src_5154, "<#5154>", "exec")
```
Comment (youknowone) [inline extra_tests/snippets/regression_closed_issues.py:145]: extra_tests/snippets/syntax_<something> will be the right place.

**Review by youknowone — CHANGES_REQUESTED** [review]

**Comment by changjoon-park** [comment]
thanks for the correct it's wrong shape
i'm closing this and will audit against Lib/test before any targeted follow-ups :)
